core/src/main/scala/kafka/tools/MirrorMaker.scala

```diff
@@ -417,22 +573,38 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
     }
   }
 
-  private class InternalRebalanceListener(connector: ZookeeperConsumerConnector,
-                                          customRebalanceListener: Option[ConsumerRebalanceListener])
-    extends ConsumerRebalanceListener {
+  private class InternalRebalanceListener(mirrorMakerConsumer: MirrorMakerBaseConsumer,
+                                          customRebalanceListenerForOldConsumer: Option[ConsumerRebalanceListener],
+                                          customRebalanceListenerForNewConsumer: Option[org.apache.kafka.clients.consumer.ConsumerRebalanceListener])
+    extends BaseConsumerRebalanceListener {
 
-    override def beforeReleasingPartitions(partitionOwnership: java.util.Map[String, java.util.Set[java.lang.Integer]]) {
+    private def flushAndCommit() {
       producer.flush()
-      commitOffsets(connector)
+      commitOffsets(mirrorMakerConsumer)
+    }
+
+    override def beforeReleasingPartitions(partitionOwnership: java.util.Map[String, java.util.Set[java.lang.Integer]]) {
+      flushAndCommit()
       // invoke custom consumer rebalance listener
-      if (customRebalanceListener.isDefined)
-        customRebalanceListener.get.beforeReleasingPartitions(partitionOwnership)
+      if (customRebalanceListenerForOldConsumer.isDefined)
+        customRebalanceListenerForOldConsumer.get.beforeReleasingPartitions(partitionOwnership)
     }
 
     override def beforeStartingFetchers(consumerId: String,
                                         partitionAssignment: java.util.Map[String, java.util.Map[java.lang.Integer, ConsumerThreadId]]) {
-      if (customRebalanceListener.isDefined)
-        customRebalanceListener.get.beforeStartingFetchers(consumerId, partitionAssignment)
+      if (customRebalanceListenerForOldConsumer.isDefined)
+        customRebalanceListenerForOldConsumer.get.beforeStartingFetchers(consumerId, partitionAssignment)
+    }
+
+    override def onPartitionsRevoked(partitions: util.Collection[TopicPartition]) {
+      flushAndCommit()
+      if (customRebalanceListenerForNewConsumer.isDefined)
+        customRebalanceListenerForNewConsumer.get.onPartitionsAssigned(partitions)
+    }
+
+    override def onPartitionsAssigned(partitions: util.Collection[TopicPartition]) {
+      if (customRebalanceListenerForNewConsumer.isDefined)
+        customRebalanceListenerForNewConsumer.get.onPartitionsAssigned(partitions)
```

Comment:
Safer and more concise:

``customRebalanceListenerForNewConsumer.foreach(_.onPartitionAssigned(partitions))```

There are many places in this PR that can benefit from this pattern.